### PR TITLE
CORS-2428: terraform: add build information to binaries

### PR DESCRIPTION
**Pull request by r4f4**
Store the hash of the go.mod in the built binary which we can then use to determine if a provider needs to be rebuild. Useful when we import pre-built binaries from a container image.

### CODE REVIEW

**Comment by r4f4** [comment]
Update: TIL that go doesn't support multiple `-ldflags` arguments (https://github.com/golang/go/issues/26849)

**Comment by r4f4** [comment]
Seems to be working 
```
2023-11-24T17:20:57.172351821Z Go compliance shim [7117] [][]: final command line arguments: "build" "-gcflags" "" "-ldflags" "-s -w -X installer.version=f28c31aee43aa7d9dd3c527e54a39435051f227f" "-o" "../../bin/darwin_amd64/terraform-provider-aws" "./vendor/github.com/hashicorp/terraform-provider-aws" 
```

File: terraform/Makefile [hunk 1]
```diff
@@ -34,15 +32,15 @@ $(GO_BUILD_TARGETS): go-build.%: bin/$(TARGET_OS_ARCH)/terraform-provider-%.zip
 $(TERRAFORM_PROVIDER_TARGETS): bin/$(TARGET_OS_ARCH)/terraform-provider-%.zip: providers/%/go.mod
 	cd providers/$*; \
 	if [ -f main.go ]; then path="."; else path=./vendor/`grep _ tools.go|awk '{ print $$2 }'|sed 's|"||g'`; fi; \
-	go build -gcflags $(GCFLAGS) -ldflags $(LDFLAGS) -o ../../bin/$(TARGET_OS_ARCH)/terraform-provider-$* "$$path" && \
+	go build -gcflags $(GCFLAGS) -ldflags "$(LDFLAGS) -X installer.version=$$(git hash-object go.mod)" -o ../../bin/$(TARGET_OS_ARCH)/terraform-provider-$* "$$path" && \
```
Comment (zaneb) [inline terraform/Makefile:35]: According to https://github.com/golang/go/wiki/GcToolchainTricks#including-build-information-in-the-executable "The string variable must exist, ..." but "There is no warning for using the wrong name".
It looks like we're relying on the fact that the ldflags are stored in the binary rather than anything the linker is actually doing with them? It would be good to mention that , at least in the commit message if a comment is too unwieldy.
A more descriptive name might be helpful for some future maintainer too.

Comment (r4f4) [reply terraform/Makefile:35]: What name would you suggest?

**Comment by patrickdillon** [comment]
I believe we had an alternate solution to solve this problem by comparing the `mod` version provided by `go version -m` and the `go.mod` file:

```
$ go version -m terraform/bin/linux_amd64/terraform-provider-aws | head -n 3
terraform/bin/linux_amd64/terraform-provider-aws: go1.20
	path	github.com/hashicorp/terraform-provider-aws
	mod	github.com/hashicorp/terraform-provider-aws	v1.60.1-0.20230615210323-24881e9f4460	

$ cat terraform/providers/aws/go.mod | head -n 5
module github.com/openshift/installer/terraform/providers/aws

go 1.20

require github.com/hashicorp/terraform-provider-aws v1.60.1-0.20230615210323-24881e9f4460 // v5.4.0
```

Is there a shortcoming to that approach (which is solved here) or is this just an alternate approach?

**Comment by r4f4** [comment]
> I believe we had an alternate solution to solve this problem by comparing the `mod` version provided by `go version -m` and the `go.mod` file:
> 
> ```
> $ go version -m terraform/bin/linux_amd64/terraform-provider-aws | head -n 3
> terraform/bin/linux_amd64/terraform-provider-aws: go1.20
> 	path	github.com/hashicorp/terraform-provider-aws
> 	mod	github.com/hashicorp/terraform-provider-aws	v1.60.1-0.20230615210323-24881e9f4460	
> 
> $ cat terraform/providers/aws/go.mod | head -n 5
> module github.com/openshift/installer/terraform/providers/aws
> 
> go 1.20
> 
> require github.com/hashicorp/terraform-provider-aws v1.60.1-0.20230615210323-24881e9f4460 // v5.4.0
> ```
> 
> Is there a shortcoming to that approach (which is solved here) or is this just an alternate approach?

Yes. See https://github.com/openshift/installer/pull/7688#issuecomment-1824739693

TL;DR: the provider module version does not change if you `replace dep => dep` anything that's not the provider itself.

**Comment by r4f4** [comment]
> This LGTM, although +1 to Zane's comment about the commit message. The work you're doing here is pretty complex, so it would be helpful to elaborate a bit more in the commit message.

Sure, I'll incorporate your suggestion. I think @zaneb also prefers something other than `installer.version` for the variable name, so I'll wait for other name suggestions before amending the commit message.

File: terraform/Makefile (hunk 1 above)
Comment (zaneb) [reply terraform/Makefile:35]: Maybe something like `main.builtGoModHash`?

**Comment by r4f4** [comment]
Update: amended commit message and changed the variable name used:
```bash
$ go version -m terraform/bin/linux_amd64/terraform-provider-aws
	build	-gcflags="all=-N -l"
	build	-ldflags=" -X main.builtGoModHash=f28c31aee43aa7d9dd3c527e54a39435051f227f"
```

**Comment by r4f4** [comment]
/retitle CORS-2428: terraform: add build information to binaries
